src/python/pants/base/specs.py

```diff
@@ -106,37 +108,27 @@ def __post_init__(self) -> None:
     def __str__(self) -> str:
         return f"{self.directory}:{self.name}"
 
-    def matching_address_families(self, address_families_dict: Dict[str, Any]) -> List[Any]:
-        return self.address_families_for_dir(address_families_dict, self.directory)
-
-    class _SingleAddressResolutionError(Exception):
-        def __init__(self, single_address_family: Any, name: str) -> None:
-            super().__init__()
-            self.single_address_family = single_address_family
-            self.name = name
+    def to_globs(self, build_patterns: Iterable[str]) -> Tuple[str, ...]:
+        return _globs_in_single_dir(self.directory, build_patterns)
 
-    def address_target_pairs_from_address_families(self, address_families: Sequence[Any]):
-        """Return the pair for the single target matching the single AddressFamily, or error.
+    def matching_address_families(
+        self, address_families_dict: Mapping[str, "AddressFamily"]
+    ) -> Tuple["AddressFamily", ...]:
+        return (_address_family_for_dir(self.directory, address_families_dict),)
 
-        :raises: :class:`SingleAddress._SingleAddressResolutionError` if no targets could be found for a
-                 :class:`SingleAddress` instance.
-        :return: list of (Address, Target) pairs with exactly one element.
-        """
+    def matching_addresses(
+        self, address_families: Sequence["AddressFamily"]
+    ) -> Sequence[Tuple[Address, TargetAdaptor]]:
         single_af = assert_single_element(address_families)
-        addr_tgt_pairs = [
-            (bfa, tgt)
-            for bfa, tgt in single_af.addressables.items()
-            if bfa.address.target_name == self.name
-        ]
-        if len(addr_tgt_pairs) == 0:
```

Comment:
Is it intentional that `SingleAddress` no longer requires a match? Did that move elsewhere?

Reply:
Yes, it moved to `build_files.py`, which was better than being split across two places, but still not great. I'm fixing this so that instead the logic lives entirely where it belongs, here.